src/h2load.cc

```diff
@@ -252,6 +261,55 @@ void rate_period_timeout_w_cb(struct ev_loop *loop, ev_timer *w, int revents) {
 } // namespace
 
 namespace {
+// Called when the duration for infinite number of requests are over
+void duration_timeout_cb(struct ev_loop *loop, ev_timer *w, int revents) {
+  auto client = static_cast<Client *>(w->data);
+
+  client->req_todo = client->req_done; // there was no finite "req_todo"
+  client->worker->stats.req_todo += client->req_todo;
+  client->req_inflight = 0;
+  client->req_left = 0;
+  client->measurement_calculation_done = true;
+  
+  if (client->worker->current_phase != Phase::DURATION_OVER) {
+    client->worker->current_phase = Phase::DURATION_OVER;
+    client->worker->stop_all_clients();
+  }
+}
+} // namespace
+
+namespace {
+// Called when the warmup duration for infinite number of requests are over
+void warmup_timeout_cb(struct ev_loop *loop, ev_timer *w, int revents) {
+  auto client = static_cast<Client *>(w->data);
+  
+  if (client->worker->current_phase != Phase::MAIN_DURATION) {
```

Comment:
So, I think we actually need each client to have its own phase variable. This is because we mutate the client stats here (even if some other client has switched to main phase, and then in various other places for the client we reference the worker phase variable, checking if we are in MAIN_DURATION. There could then be a race - e.g. we start counting due to the worker being in MAIN_DURATION but the client hasn't yet hit this CB.

Similarly, we can print the phase transition for each client.

Apologies that this contracdicts what we discussed earlier, but it's more apparent now that the code is cleaner.